metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -30,19 +32,38 @@ import scala.annotation.tailrec
 
 class CheckpointingService[F[_]: Sync, N](
     ledgerTreeRef: Ref[F, LedgerTree],
+    mempoolRef: Ref[F, Mempool],
     lastCommittedHeaderRef: Ref[F, Block.Header],
     checkpointDataRef: Ref[F, Option[CheckpointData]],
-    //TODO: PM-3137, this is used for testing that a certificate was created correctly
-    //      replace with proper means of pushing the certificate to the Interpreter
-    pushCheckpointFn: CheckpointCertificate => F[Unit],
     ledgerStorage: LedgerStorage[N],
-    blockStorage: BlockStorage[N, CheckpointingAgreement]
+    blockStorage: BlockStorage[N, CheckpointingAgreement],
+    interpreterClient: InterpreterRPC[F]
 )(implicit storeRunner: KVStoreRunner[F, N])
     extends ApplicationService[F, CheckpointingAgreement] {
 
   override def createBlock(
       highQC: QuorumCertificate[CheckpointingAgreement]
-  ): F[Option[Block]] = ???
+  ): F[Option[Block]] = (
+    for {
+      parent    <- OptionT(getBlock(highQC.blockHash))
+      oldLedger <- OptionT(projectLedger(parent))
+      mempool   <- OptionT.liftF(projectMempool)
+
+      newBody <- OptionT {
+        if (mempool.isEmpty)
```

Comment:
It may be worth making this behaviour configurable, in case the interpreter doesn't support notifications for whatever reason.

Reply:
Yeah, makes sense, but do you mean to include this config now or just make a note of it? Asking, cause I don't believe we actually have any config settings yet. Do you have any concrete vision as to how they should be structured?

Reply:
What I do in these cases is either:
* Add a constructor argument to indicate that this is an optional behaviour that I can tweak later during application composition.
* When I have two or more of these options, I usually create `case class Config` in the companion object to hold all possible options for this class, perhaps with a `val default = Config(...)` for the default options that I can use as a default parameter value in `apply` and pass to the constructor. For example [ConsensusService.Config](https://github.com/input-output-hk/metronome/blob/pm-3353-quorum-certificate-phase-typed/metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/ConsensusService.scala#L580), which gets passed down through `HotStuffService`.

This second option can then be naturally used by the `config` module, so we can have a root config object like this:
```scala
case class CheckpointingConfig(
  network: CheckpointingConfig.Network, 
  checkpointing: CheckpointingService.Config
)
```
Or something along those lines. [RobotConfig](https://github.com/input-output-hk/metronome/pull/42/files#diff-1495f70cdcd12e0a2f23837ea37dd12921eba8d68262c606737f046fb9a0e123) is an example, used throughout [RobotComposition](https://github.com/input-output-hk/metronome/pull/42/files#diff-10b56ac2536b604ca590b5fe6859da38983d0023e5199eb3d5659d1d9cee12ddR72-R75) to pass the values to things that demand it.

Reply:
Oh, thanks, I did not realise we already have something like that. Yeah I would definitely prefer to accumulate those settings using a case class as there will certainly be more. Interpreter timeout being one example. 

I was also thinking that there should be a setting whether to expect any proposer blocks. In case of ETC (`false`) it wouldn't make sense to filter out mempool, especially that it may need to read blocks from storage.